prov/shm/src/smr_dsa.c

```diff
@@ -148,16 +150,25 @@ static inline unsigned char dsa_enqcmd(struct dsa_hw_desc *desc,
 static __always_inline void dsa_desc_submit(struct smr_dsa_context *dsa_context,
 					    struct dsa_hw_desc *hw)
 {
-	int enq_status;
+	int ret, enq_status;
 
 	// make sure writes (e.g., comp.status = 0) are ordered wrt to enqcmd
 	{ asm volatile("sfence":::"memory"); }
 
+	//TODO: Should we set a max retry limit?
 	do {
-		enq_status = dsa_enqcmd(hw,
+		if (dsa_context->wq_mmap_support)
+			enq_status = dsa_enqcmd(
+				hw,
 				dsa_context->wq_portal[dsa_context->next_wq]);
-		dsa_context->next_wq = (dsa_context->next_wq + 1) %
-			(dsa_context->wq_count);
+		else {
```

Comment:
else if wq_write_supported. The the final else is an assert since as of now we  should not enter this flow if either one is not supported.